### PR TITLE
[SYCL] Add E2E ESIMD bit operations test.

### PR DESCRIPTION
**Pull request by kbobrovs**
The feature: https://github.com/intel/llvm/pull/3847

Signed-off-by: kbobrovs <Konstantin.S.Bobrovsky@intel.com>

### CODE REVIEW

File: SYCL/ESIMD/api/esimd_bit_ops.cpp [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+  T val_all_one{ static_cast<T>(~val_all_zero) };
+  T val_two_one{ static_cast<T>(T{1} << (sizeof(T) * 8 - 2) | 2) }; // 010...010
+
+  T vals[] = { val_all_zero, val_all_one, val_two_one };
+  constexpr size_t num_vals = sizeof(vals) / sizeof(vals[0]);
+
+  constexpr size_t size = N * num_vals;
+  unsigned int *A = sycl::malloc_shared<unsigned int>(size, q);
+
+  for (unsigned int i = 0; i < size; ++i) {
```
Comment (DenisBakhvalov) [inline SYCL/ESIMD/api/esimd_bit_ops.cpp]:
```suggestion
  for (unsigned int i = 0; i < num_vals; ++i) {
```

File: SYCL/ESIMD/api/esimd_bit_ops.cpp [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+
+template <typename T> struct char_to_int {
+  using type = typename std::conditional<
+      sizeof(T) == 1,
+      typename std::conditional<std::is_signed<T>::value, int, unsigned>::type,
+      T>::type;
+};
+
+// The main test routine.
+template <typename T, int N, int Op> bool test(queue q) {
```
Comment (DenisBakhvalov) [inline SYCL/ESIMD/api/esimd_bit_ops.cpp:39]: Maybe pick a more verbose name for `N`?
```suggestion
template <typename T, int Nbits, int Op> bool test(queue q) {
```

Comment (kbobrovs) [reply SYCL/ESIMD/api/esimd_bit_ops.cpp:39]: N is not Nbits, it is the number of elements in ESIMD vector, as used in lots of other places.

File: SYCL/ESIMD/api/esimd_bit_ops.cpp (hunk 1 above)
Comment (kbobrovs) [reply SYCL/ESIMD/api/esimd_bit_ops.cpp]: why? I want to initialize the entire array.

Comment (kbobrovs) [reply SYCL/ESIMD/api/esimd_bit_ops.cpp]: actually, I should allocate fewer elements too, will update (remnant from the previous version of the test)

File: SYCL/ESIMD/api/esimd_bit_ops.cpp [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+
+  T val_all_zero{ 0 };
+  T val_all_one{ static_cast<T>(~val_all_zero) };
+  T val_two_one{ static_cast<T>(T{1} << (sizeof(T) * 8 - 2) | 2) }; // 010...010
+
+  T vals[] = { val_all_zero, val_all_one, val_two_one };
+  constexpr size_t num_vals = sizeof(vals) / sizeof(vals[0]);
+
+  constexpr size_t size = N * num_vals;
+  unsigned int *A = sycl::malloc_shared<unsigned int>(size, q);
```
Comment (kbobrovs) [inline SYCL/ESIMD/api/esimd_bit_ops.cpp]:
```suggestion
  unsigned int *A = sycl::malloc_shared<unsigned int>(num_vals, q);
```

**Review by DenisBakhvalov — DISMISSED** [review]

File: SYCL/ESIMD/api/esimd_bit_ops.cpp [hunk 4]
```diff
@@ -0,0 +1,146 @@
+//==------- esimd_bit_ops.cpp  - DPC++ ESIMD on-device test ----------------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+// REQUIRES: gpu
+// UNSUPPORTED: cuda
+// RUN: %clangxx -fsycl %s -fsycl-device-code-split=per_kernel -o %t.out
```
Comment (vladimirlaz) [inline SYCL/ESIMD/api/esimd_bit_ops.cpp:10]:
```suggestion
// RUN: %clangxx -fsycl %s -o %t.out
```
Is the option mandatory? if not it is better to remove it. To let us run testing with different values of this option.

Comment (kbobrovs) [reply SYCL/ESIMD/api/esimd_bit_ops.cpp:10]: Yes, this option is needed to test all cases even if some lead to BE compilation failure.

**Review by DenisBakhvalov — APPROVED** [review]

**Pull request merged**